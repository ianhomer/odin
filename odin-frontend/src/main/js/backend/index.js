// Copyright (c) 2017 The Odin Authors. All Rights Reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.



import {call, put, takeEvery, getContext} from 'redux-saga/effects'
import fetch from 'isomorphic-fetch'
import {
  CREATE_ENTITY_REQUESTED, CREATE_ENTITY_SUCCEEDED, CREATE_ENTITY_FAILED,
  UPDATE_ENTITY_REQUESTED, UPDATE_ENTITY_SUCCEEDED, UPDATE_ENTITY_FAILED,
  DELETE_ENTITY_REQUESTED, DELETE_ENTITY_SUCCEEDED, DELETE_ENTITY_FAILED,
  PATCH_ENTITY_REQUESTED, PATCH_ENTITY_SUCCEEDED, PATCH_ENTITY_FAILED,
  FETCH_COMPOSITION_REQUESTED, FETCH_COMPOSITION_SUCCEEDED, FETCH_COMPOSITION_FAILED,
  FETCH_SYSTEM_REQUESTED, FETCH_SYSTEM_SUCCEEDED, FETCH_SYSTEM_FAILED,
  LOAD_ENTITIES_REQUESTED, LOAD_ENTITIES_SUCCEEDED, LOAD_ENTITIES_FAILED,
  LOAD_PERFORMANCE_SCHEMA_REQUESTED, LOAD_PERFORMANCE_SCHEMA_SUCCEEDED, LOAD_PERFORMANCE_SCHEMA_FAILED,
  LOAD_PROFILE_SCHEMA_REQUESTED, LOAD_PROFILE_SCHEMA_SUCCEEDED, LOAD_PROFILE_SCHEMA_FAILED
} from '../actions'
import {withQuery} from './withQuery'

const apiRoot = '/api'
const restRoot = apiRoot + '/rest'

export class Backend {
  createEntityApi(entity, path) {
    return fetch(restRoot + '/' + path, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      }, body: JSON.stringify(entity)
    })
      .then(response => response.json())
      .catch(reason => console.error(reason))
  }

  *createEntity(action) {
    try {
      const backend = yield getContext('backend')
      const entity = yield call(backend.createEntityApi, action.entity, action.path)
      yield put({type: CREATE_ENTITY_SUCCEEDED, entity: entity, path: action.path})
      yield call(backend.enrichEntity, entity)
    } catch (e) {
      yield put({type: CREATE_ENTITY_FAILED, message: e.message})
    }
  }

  updateEntityApi(entity) {
    return fetch(entity._links.self.href, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      }, body: JSON.stringify(entity)
    })
      .then(response => response.json())
      .catch(reason => console.error(reason))
  }

  *updateEntity(action) {
    try {
      const backend = yield getContext('backend')
      const entity = yield call(backend.updateEntityApi, action.entity)
      yield call(backend.enrichEntity, entity)
      yield put({type: UPDATE_ENTITY_SUCCEEDED, entity: entity, path: action.path})
    } catch (e) {
      yield put({type: UPDATE_ENTITY_FAILED, message: e.message})
    }
  }

  fetchCompositionApi(notation) {
    return fetch(withQuery(apiRoot + '/services/composition', {notation}), {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json'
      }
    })
      .then(response => response.json())
      .catch(reason => console.error(reason))
  }

  *fetchComposition(action) {
    try {
      const backend = yield getContext('backend')
      const composition = yield call(backend.fetchCompositionApi, action.notation)
      yield put({type: FETCH_COMPOSITION_SUCCEEDED, entityName: action.entityName,
        notation: action.notation, composition})
    } catch (e) {
      yield put({type: FETCH_COMPOSITION_FAILED, message: e.message})
    }
  }

  fetchSystemApi(path) {
    return fetch(apiRoot + '/services/system/' + path, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json'
      }
    })
      .then(response => response.json())
      .catch(reason => console.error(reason))
  }

  *fetchSystem(action) {
    try {
      const backend = yield getContext('backend')
      const system = yield call(backend.fetchSystemApi, action.path)
      yield put({type: FETCH_SYSTEM_SUCCEEDED, path: action.path, system})
    } catch (e) {
      yield put({type: FETCH_SYSTEM_FAILED, message: e.message})
    }
  }

  *enrichEntity(entity) {
    if (entity.type === 'notation') {
      yield put({type: FETCH_COMPOSITION_REQUESTED, entityName: entity.name, notation: entity.properties.notation})
    }
  }

  deleteEntityApi(entity) {
    return fetch(entity._links.self.href, {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json'
      }
    })
      .then(() => entity)
      .catch(reason => console.error(reason))
  }

  *deleteEntity(action) {
    try {
      const backend = yield getContext('backend')
      const entity = yield call(backend.deleteEntityApi, action.entity)
      yield put({type: DELETE_ENTITY_SUCCEEDED, entity: entity, path: action.path})
    } catch (e) {
      yield put({type: DELETE_ENTITY_FAILED, message: e.message})
    }
  }

  patchEntityApi(entity, patch) {
    return fetch(entity._links.self.href, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json-patch+json'
        // TODO : Etag support, note that entity needs to be loaded from server prior to
        // editing to populate Etag
        //'If-Match': entity.headers.Etag
      },
      body: JSON.stringify(patch)
    })
      .then(response => response.json())
      .catch(reason => console.error(reason))
  }

  *patchEntity(action) {
    try {
      const backend = yield getContext('backend')
      const entity = yield call(backend.patchEntityApi, action.entity, action.patch)
      yield put({type: PATCH_ENTITY_SUCCEEDED, entity: entity, path: action.path})
      // Support chain of patches
      if (action.nextAction) {
        yield put(action.nextAction)
      }
    } catch (e) {
      yield put({type: PATCH_ENTITY_FAILED, message: e.message})
    }
  }

  loadEntitiesApi(path) {
    return fetch(restRoot + '/' + path, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json'
      }
    })
      .then(response => response.json())
      .then(json => json._embedded[path])
      .catch(reason => console.error(reason))
  }

  *loadEntities(action) {
    try {
      const backend = yield getContext('backend')
      const entities = yield call(backend.loadEntitiesApi, action.path, action.schema)
      for (var i = 0; i < entities.length ; i++) {
        yield call(backend.enrichEntity, entities[i])
        // Set the path value in the entity so that the front end knows the type of entity.
        entities[i].path = action.path
      }
      yield put({type: LOAD_ENTITIES_SUCCEEDED, entities: entities, path: action.path})
    } catch (e) {
      yield put({type: LOAD_ENTITIES_FAILED, message: e.message})
    }
  }

  loadPerformanceSchemaApi() {
    return fetch('/api/services/schema', {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json'
      }
    })
      .then(response => response.json())
      .catch(reason => console.error(reason))
  }

  *loadPerformanceSchema() {
    try {
      const backend = yield getContext('backend')
      const schema = yield call(backend.loadPerformanceSchemaApi)
      yield put({type: LOAD_PERFORMANCE_SCHEMA_SUCCEEDED, schema: schema})
    } catch (e) {
      yield put({type: LOAD_PERFORMANCE_SCHEMA_FAILED, message: e.message})
    }
  }

  loadProfileSchemaApi(path) {
    return fetch(restRoot + '/profile/' + path, {
      method: 'GET',
      headers: {
        'Accept': 'application/schema+json'
      }
    })
      .then(response => response.json())
      .catch(reason => console.error(reason))
  }

  *loadProfileSchema(action) {
    try {
      const backend = yield getContext('backend')
      const schema = yield call(backend.loadProfileSchemaApi, action.path)
      yield put({type: LOAD_PROFILE_SCHEMA_SUCCEEDED, schema: schema, path: action.path})
    } catch (e) {
      yield put({type: LOAD_PROFILE_SCHEMA_FAILED, message: e.message})
    }
  }

  *saga() {
    var backend = yield getContext('backend')
    yield takeEvery(CREATE_ENTITY_REQUESTED, backend.createEntity)
    yield takeEvery(UPDATE_ENTITY_REQUESTED, backend.updateEntity)
    yield takeEvery(DELETE_ENTITY_REQUESTED, backend.deleteEntity)
    yield takeEvery(PATCH_ENTITY_REQUESTED, backend.patchEntity)
    yield takeEvery(FETCH_COMPOSITION_REQUESTED, backend.fetchComposition)
    yield takeEvery(FETCH_SYSTEM_REQUESTED, backend.fetchSystem)
    yield takeEvery(LOAD_ENTITIES_REQUESTED, backend.loadEntities)
    yield takeEvery(LOAD_PERFORMANCE_SCHEMA_REQUESTED, backend.loadPerformanceSchema)
    yield takeEvery(LOAD_PROFILE_SCHEMA_REQUESTED, backend.loadProfileSchema)
  }
}
