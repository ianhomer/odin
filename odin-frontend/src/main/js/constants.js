export const ItemTypes = {
  LAYER: 'layer',
  SEQUENCE: 'sequence',
  SEQUENCE_IN_LAYER: 'sequenceInLayer'
}

// If global variable odinApiHost is set then use that as the host, otherwise we expect it's local
/* global odinApiHost */
export const apiRoot = (typeof odinApiHost !== 'undefined' ? odinApiHost : '') + '/api'