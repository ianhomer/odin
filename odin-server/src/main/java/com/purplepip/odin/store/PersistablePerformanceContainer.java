/*
 * Copyright (c) 2017 the original author or authors. All Rights Reserved
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.purplepip.odin.store;

import com.purplepip.odin.creation.channel.Channel;
import com.purplepip.odin.creation.layer.MutableLayer;
import com.purplepip.odin.creation.sequence.SequenceConfiguration;
import com.purplepip.odin.performance.PerformanceContainer;
import com.purplepip.odin.server.rest.repositories.ChannelRepository;
import com.purplepip.odin.server.rest.repositories.LayerRepository;
import com.purplepip.odin.server.rest.repositories.SequenceRepository;
import com.purplepip.odin.store.domain.PersistableChannel;
import com.purplepip.odin.store.domain.PersistableLayer;
import com.purplepip.odin.store.domain.PersistableSequence;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
@Slf4j
public class PersistablePerformanceContainer extends PerformanceContainer {
  @Autowired
  private LayerRepository layerRepository;

  @Autowired
  private ChannelRepository channelRepository;

  @Autowired
  private SequenceRepository sequenceRepository;

  @Override
  public void addChannel(Channel channel) {
    if (channel instanceof PersistableChannel) {
      ((PersistableChannel) channel).setPerformance(getPerformance());
      channelRepository.save((PersistableChannel) channel);
    }
    super.addChannel(channel);
  }

  @Override
  public PersistablePerformanceContainer addSequence(SequenceConfiguration sequence) {
    LOG.debug("Adding sequence : {}", sequence);
    if (sequence instanceof PersistableSequence) {
      PersistableSequence persistableSequence = ((PersistableSequence) sequence);
      LOG.debug("Saving persistable sequence : {}", sequence);
      persistableSequence.setPerformance(getPerformance());
      sequenceRepository.save((PersistableSequence) sequence);
    }
    super.addSequence(sequence);
    return this;
  }

  @Override
  public PersistablePerformanceContainer addLayer(MutableLayer layer) {
    if (layer instanceof PersistableLayer) {
      ((PersistableLayer) layer).setPerformance(getPerformance());
      layerRepository.save((PersistableLayer) layer);
    }
    super.addLayer(layer);
    return this;
  }
  // TODO : Create addTrigger implementation
}
