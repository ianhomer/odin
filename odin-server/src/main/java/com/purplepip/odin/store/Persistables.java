/*
 * Copyright (c) 2017 Ian Homer. All Rights Reserved
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.purplepip.odin.store;

import com.purplepip.odin.store.domain.PersistableThing;

/**
 * Persistable utility functions.
 */
public final class Persistables {
  /**
   * Sometimes toSting() method on an entity is not safe, since the entity might have lazy
   * collections that have not been initialised.  An example of this situation is trying to
   * log the entity toString() after it has been deleted from the store.
   *
   * @param object object to output as string.
   * @return string representation
   */
  public static String toThingString(Object object) {
    String string;
    if (object instanceof PersistableThing) {
      /*
       * Just output the thing string which know will be safe since it does not have any
       * lazy properties.
       */
      string = ((PersistableThing) object).toThingString();
    } else {
      /*
       * If it's not a persistable thing, just call toString as normal.
       */
      string = object.toString();
    }
    return string;
  }

  private Persistables() {
  }
}
