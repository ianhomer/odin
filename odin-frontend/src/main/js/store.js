// Copyright (c) 2017 The Odin Authors. All Rights Reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import thunkMiddleware from 'redux-thunk'
import {createLogger} from 'redux-logger'
import {applyMiddleware, combineReducers, createStore} from 'redux'
import createSagaMiddleware from 'redux-saga'

import reducers from './reducers'
import {Backend} from './backend'
import {loadSchemaActions} from './actions'

// Create the backend integration layer
const backend = new Backend()

// Create saga middleware for handling of calls to backend.  Note that the backend is
// passed in as context for the saga middleware so that we can inject the backend API calls in.
// Injecting backend API calls in is required for switching in different backends as required
// for test cases which need a mock backend.
const sagaMiddleware = createSagaMiddleware({context: {backend: backend}})

const middleware = [thunkMiddleware, sagaMiddleware]
const env = process.env.NODE_ENV || 'dev'
if (env === 'dev') {
  middleware.push(createLogger({collapsed: true}))
}

// Create the application store
const store = createStore(
  combineReducers(reducers),
  applyMiddleware(...middleware)
)

// Hook in back end API calls
sagaMiddleware.run(backend.saga)

for (let i = 0; i < loadSchemaActions.length ; i++) {
  store.dispatch(loadSchemaActions[i])
}

export default store
