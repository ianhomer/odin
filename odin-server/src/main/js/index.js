// Copyright (c) 2017 the original author or authors. All Rights Reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react'
import { render } from 'react-dom'

import thunkMiddleware from 'redux-thunk'
import { createLogger } from 'redux-logger'
import { applyMiddleware, combineReducers, createStore } from 'redux'
import 'babel-polyfill'
import createSagaMiddleware from 'redux-saga'

import App from './containers/app'
import reducers from './reducers'
import { Flux } from './legacy/flux'
import { Backend } from './backend'
import { loadSchemaActions } from './actions'

// Create the backend integration layer
const backend = new Backend()

// Create saga middleware for handling of calls to backend.  Note that the backend is
// passed in as context for the saga middleware so that we can inject the backend API calls in.
// Injecting backend API calls in is required for switching in different backends as required
// for test cases which need a mock backend.
const sagaMiddleware = createSagaMiddleware({context: {backend: backend}})

// Create the application store
const store = createStore(
  combineReducers(reducers),
  applyMiddleware(
    thunkMiddleware,
    sagaMiddleware,
    createLogger({collapsed: true})
  )
)

// Hook in back end API calls
sagaMiddleware.run(backend.saga)

for (var i = 0; i < loadSchemaActions.length ; i++) {
  store.dispatch(loadSchemaActions[i])
}

// Render application
render(
  <App store={store} flux={new Flux()}/>,
  document.getElementById('root')
)