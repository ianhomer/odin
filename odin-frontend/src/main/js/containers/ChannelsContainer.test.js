import React from 'react'
import ChannelsContainer from './ChannelsContainer.js'
import {Provider} from 'react-redux'
import {mount} from '../../../test/js/enzyme.js'

import {testPerformance, testSchema} from '../../../test/js/testData.js'
import store from '../../../test/js/store'

describe('Channels container', () => {

  test('Initialise', () => {
    const component = mount(
      <Provider store={store}>
        <ChannelsContainer schema={testSchema} performance={testPerformance}/>
      </Provider>
    )
    expect(component).toMatchSnapshot()
    expect(component.find('button')).toHaveLength(6)
    component.find('button').at(0).simulate('click')
    expect(component.find('button')).toHaveLength(5)
    component.find('button').at(0).simulate('click')
    expect(component.find('button')).toHaveLength(4)
  })
})