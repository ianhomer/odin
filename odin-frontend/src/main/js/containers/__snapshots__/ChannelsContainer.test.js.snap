// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channels container Initialise 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(ChannelsContainer)
    performance={
      Object {
        "_embedded": Object {
          "performance": Array [
            Object {
              "_links": Object {
                "channels": Object {
                  "href": "http://localhost/api/rest/performance/34/channels",
                },
                "layers": Object {
                  "href": "http://localhost/api/rest/performance/34/layers",
                },
                "performance": Object {
                  "href": "http://localhost/api/rest/performance/34",
                },
                "self": Object {
                  "href": "http://localhost/api/rest/performance/34",
                },
                "sequences": Object {
                  "href": "http://localhost/api/rest/performance/34/sequences",
                },
                "triggers": Object {
                  "href": "http://localhost/api/rest/performance/34/triggers",
                },
              },
              "name": "com/purplepip/odin/demo/GroovePerformance",
            },
          ],
        },
        "_links": Object {
          "profile": Object {
            "href": "http://localhost/api/rest/profile/performance",
          },
          "search": Object {
            "href": "http://localhost/api/rest/performance/search",
          },
          "self": Object {
            "href": "http://localhost/api/rest/performance",
          },
        },
      }
    }
    schema={
      Schema {
        "revision": undefined,
      }
    }
  >
    <ChannelsContainer
      collection={
        Object {
          "entities": Array [
            Object {
              "_links": Object {
                "channel": Object {
                  "href": "http://localhost/api/rest/channel/36",
                },
                "performance": Object {
                  "href": "http://localhost/api/rest/channel/36/performance",
                },
                "self": Object {
                  "href": "http://localhost/api/rest/channel/36",
                },
              },
              "name": "Power Drums",
              "number": 9,
              "path": "channel",
              "program": 0,
              "programName": "Power Drums",
            },
            Object {
              "_links": Object {
                "channel": Object {
                  "href": "http://localhost/api/rest/channel/37",
                },
                "performance": Object {
                  "href": "http://localhost/api/rest/channel/37/performance",
                },
                "self": Object {
                  "href": "http://localhost/api/rest/channel/37",
                },
              },
              "name": "aahs",
              "number": 4,
              "path": "channel",
              "program": 0,
              "programName": "aahs",
            },
            Object {
              "_links": Object {
                "channel": Object {
                  "href": "http://localhost/api/rest/channel/38",
                },
                "performance": Object {
                  "href": "http://localhost/api/rest/channel/38/performance",
                },
                "self": Object {
                  "href": "http://localhost/api/rest/channel/38",
                },
              },
              "name": "bass",
              "number": 5,
              "path": "channel",
              "program": 0,
              "programName": "bass",
            },
            Object {
              "_links": Object {
                "channel": Object {
                  "href": "http://localhost/api/rest/channel/35",
                },
                "performance": Object {
                  "href": "http://localhost/api/rest/channel/35/performance",
                },
                "self": Object {
                  "href": "http://localhost/api/rest/channel/35",
                },
              },
              "name": "piano",
              "number": 1,
              "path": "channel",
              "program": 0,
              "programName": "piano",
            },
            Object {
              "_links": Object {
                "channel": Object {
                  "href": "http://localhost/api/rest/channel/39",
                },
                "performance": Object {
                  "href": "http://localhost/api/rest/channel/39/performance",
                },
                "self": Object {
                  "href": "http://localhost/api/rest/channel/39",
                },
              },
              "name": "rock",
              "number": 2,
              "path": "channel",
              "program": 0,
              "programName": "rock",
            },
            Object {
              "_links": Object {
                "channel": Object {
                  "href": "http://localhost/api/rest/channel/40",
                },
                "performance": Object {
                  "href": "http://localhost/api/rest/channel/40/performance",
                },
                "self": Object {
                  "href": "http://localhost/api/rest/channel/40",
                },
              },
              "name": "strings",
              "number": 3,
              "path": "channel",
              "program": 0,
              "programName": "strings",
            },
          ],
          "newEntities": Object {},
        }
      }
      onCreate={[Function]}
      onDelete={[Function]}
      onLoadEntities={[Function]}
      onUpdate={[Function]}
      path="channel"
      performance={
        Object {
          "_embedded": Object {
            "performance": Array [
              Object {
                "_links": Object {
                  "channels": Object {
                    "href": "http://localhost/api/rest/performance/34/channels",
                  },
                  "layers": Object {
                    "href": "http://localhost/api/rest/performance/34/layers",
                  },
                  "performance": Object {
                    "href": "http://localhost/api/rest/performance/34",
                  },
                  "self": Object {
                    "href": "http://localhost/api/rest/performance/34",
                  },
                  "sequences": Object {
                    "href": "http://localhost/api/rest/performance/34/sequences",
                  },
                  "triggers": Object {
                    "href": "http://localhost/api/rest/performance/34/triggers",
                  },
                },
                "name": "com/purplepip/odin/demo/GroovePerformance",
              },
            ],
          },
          "_links": Object {
            "profile": Object {
              "href": "http://localhost/api/rest/profile/performance",
            },
            "search": Object {
              "href": "http://localhost/api/rest/performance/search",
            },
            "self": Object {
              "href": "http://localhost/api/rest/performance",
            },
          },
        }
      }
      schema={
        Schema {
          "revision": undefined,
        }
      }
    >
      <div>
        <ChannelList
          collection={
            Object {
              "entities": Array [
                Object {
                  "_links": Object {
                    "channel": Object {
                      "href": "http://localhost/api/rest/channel/36",
                    },
                    "performance": Object {
                      "href": "http://localhost/api/rest/channel/36/performance",
                    },
                    "self": Object {
                      "href": "http://localhost/api/rest/channel/36",
                    },
                  },
                  "name": "Power Drums",
                  "number": 9,
                  "path": "channel",
                  "program": 0,
                  "programName": "Power Drums",
                },
                Object {
                  "_links": Object {
                    "channel": Object {
                      "href": "http://localhost/api/rest/channel/37",
                    },
                    "performance": Object {
                      "href": "http://localhost/api/rest/channel/37/performance",
                    },
                    "self": Object {
                      "href": "http://localhost/api/rest/channel/37",
                    },
                  },
                  "name": "aahs",
                  "number": 4,
                  "path": "channel",
                  "program": 0,
                  "programName": "aahs",
                },
                Object {
                  "_links": Object {
                    "channel": Object {
                      "href": "http://localhost/api/rest/channel/38",
                    },
                    "performance": Object {
                      "href": "http://localhost/api/rest/channel/38/performance",
                    },
                    "self": Object {
                      "href": "http://localhost/api/rest/channel/38",
                    },
                  },
                  "name": "bass",
                  "number": 5,
                  "path": "channel",
                  "program": 0,
                  "programName": "bass",
                },
                Object {
                  "_links": Object {
                    "channel": Object {
                      "href": "http://localhost/api/rest/channel/35",
                    },
                    "performance": Object {
                      "href": "http://localhost/api/rest/channel/35/performance",
                    },
                    "self": Object {
                      "href": "http://localhost/api/rest/channel/35",
                    },
                  },
                  "name": "piano",
                  "number": 1,
                  "path": "channel",
                  "program": 0,
                  "programName": "piano",
                },
                Object {
                  "_links": Object {
                    "channel": Object {
                      "href": "http://localhost/api/rest/channel/39",
                    },
                    "performance": Object {
                      "href": "http://localhost/api/rest/channel/39/performance",
                    },
                    "self": Object {
                      "href": "http://localhost/api/rest/channel/39",
                    },
                  },
                  "name": "rock",
                  "number": 2,
                  "path": "channel",
                  "program": 0,
                  "programName": "rock",
                },
                Object {
                  "_links": Object {
                    "channel": Object {
                      "href": "http://localhost/api/rest/channel/40",
                    },
                    "performance": Object {
                      "href": "http://localhost/api/rest/channel/40/performance",
                    },
                    "self": Object {
                      "href": "http://localhost/api/rest/channel/40",
                    },
                  },
                  "name": "strings",
                  "number": 3,
                  "path": "channel",
                  "program": 0,
                  "programName": "strings",
                },
              ],
              "newEntities": Object {},
            }
          }
          fields={
            Object {
              "number": Object {
                "cellWidth": 2,
                "defaultValue": 1,
              },
              "program": Object {
                "defaultValue": 0,
                "hidden": true,
              },
              "programName": Object {
                "cellWidth": 3,
                "defaultValue": "piano",
              },
            }
          }
          onCreate={[Function]}
          onDelete={[Function]}
          onUpdate={[Function]}
          path="channel"
          performance={
            Object {
              "_embedded": Object {
                "performance": Array [
                  Object {
                    "_links": Object {
                      "channels": Object {
                        "href": "http://localhost/api/rest/performance/34/channels",
                      },
                      "layers": Object {
                        "href": "http://localhost/api/rest/performance/34/layers",
                      },
                      "performance": Object {
                        "href": "http://localhost/api/rest/performance/34",
                      },
                      "self": Object {
                        "href": "http://localhost/api/rest/performance/34",
                      },
                      "sequences": Object {
                        "href": "http://localhost/api/rest/performance/34/sequences",
                      },
                      "triggers": Object {
                        "href": "http://localhost/api/rest/performance/34/triggers",
                      },
                    },
                    "name": "com/purplepip/odin/demo/GroovePerformance",
                  },
                ],
              },
              "_links": Object {
                "profile": Object {
                  "href": "http://localhost/api/rest/profile/performance",
                },
                "search": Object {
                  "href": "http://localhost/api/rest/performance/search",
                },
                "self": Object {
                  "href": "http://localhost/api/rest/performance",
                },
              },
            }
          }
          schema={
            Schema {
              "revision": undefined,
            }
          }
        >
          <div>
            <div
              className="container"
            >
              <div
                className="row"
              >
                <div
                  className="col-2"
                >
                  Number
                </div>
                <div
                  className="col-3"
                >
                  Program
                </div>
              </div>
              <Channel
                entity={
                  Object {
                    "_links": Object {
                      "channel": Object {
                        "href": "http://localhost/api/rest/channel/36",
                      },
                      "performance": Object {
                        "href": "http://localhost/api/rest/channel/36/performance",
                      },
                      "self": Object {
                        "href": "http://localhost/api/rest/channel/36",
                      },
                    },
                    "name": "Power Drums",
                    "number": 9,
                    "path": "channel",
                    "program": 0,
                    "programName": "Power Drums",
                  }
                }
                key="9"
                onDelete={[Function]}
              >
                <div
                  className="row"
                >
                  <div
                    className="col-2"
                  >
                    9
                  </div>
                  <div
                    className="col-3"
                  >
                    Power Drums
                  </div>
                  <div
                    className="col-1"
                  >
                    <button
                      className="btn btn-primary"
                      onClick={[Function]}
                      type="submit"
                    >
                      Delete
                    </button>
                  </div>
                </div>
              </Channel>
              <Channel
                entity={
                  Object {
                    "_links": Object {
                      "channel": Object {
                        "href": "http://localhost/api/rest/channel/37",
                      },
                      "performance": Object {
                        "href": "http://localhost/api/rest/channel/37/performance",
                      },
                      "self": Object {
                        "href": "http://localhost/api/rest/channel/37",
                      },
                    },
                    "name": "aahs",
                    "number": 4,
                    "path": "channel",
                    "program": 0,
                    "programName": "aahs",
                  }
                }
                key="4"
                onDelete={[Function]}
              >
                <div
                  className="row"
                >
                  <div
                    className="col-2"
                  >
                    4
                  </div>
                  <div
                    className="col-3"
                  >
                    aahs
                  </div>
                  <div
                    className="col-1"
                  >
                    <button
                      className="btn btn-primary"
                      onClick={[Function]}
                      type="submit"
                    >
                      Delete
                    </button>
                  </div>
                </div>
              </Channel>
              <Channel
                entity={
                  Object {
                    "_links": Object {
                      "channel": Object {
                        "href": "http://localhost/api/rest/channel/38",
                      },
                      "performance": Object {
                        "href": "http://localhost/api/rest/channel/38/performance",
                      },
                      "self": Object {
                        "href": "http://localhost/api/rest/channel/38",
                      },
                    },
                    "name": "bass",
                    "number": 5,
                    "path": "channel",
                    "program": 0,
                    "programName": "bass",
                  }
                }
                key="5"
                onDelete={[Function]}
              >
                <div
                  className="row"
                >
                  <div
                    className="col-2"
                  >
                    5
                  </div>
                  <div
                    className="col-3"
                  >
                    bass
                  </div>
                  <div
                    className="col-1"
                  >
                    <button
                      className="btn btn-primary"
                      onClick={[Function]}
                      type="submit"
                    >
                      Delete
                    </button>
                  </div>
                </div>
              </Channel>
              <Channel
                entity={
                  Object {
                    "_links": Object {
                      "channel": Object {
                        "href": "http://localhost/api/rest/channel/35",
                      },
                      "performance": Object {
                        "href": "http://localhost/api/rest/channel/35/performance",
                      },
                      "self": Object {
                        "href": "http://localhost/api/rest/channel/35",
                      },
                    },
                    "name": "piano",
                    "number": 1,
                    "path": "channel",
                    "program": 0,
                    "programName": "piano",
                  }
                }
                key="1"
                onDelete={[Function]}
              >
                <div
                  className="row"
                >
                  <div
                    className="col-2"
                  >
                    1
                  </div>
                  <div
                    className="col-3"
                  >
                    piano
                  </div>
                  <div
                    className="col-1"
                  >
                    <button
                      className="btn btn-primary"
                      onClick={[Function]}
                      type="submit"
                    >
                      Delete
                    </button>
                  </div>
                </div>
              </Channel>
              <Channel
                entity={
                  Object {
                    "_links": Object {
                      "channel": Object {
                        "href": "http://localhost/api/rest/channel/39",
                      },
                      "performance": Object {
                        "href": "http://localhost/api/rest/channel/39/performance",
                      },
                      "self": Object {
                        "href": "http://localhost/api/rest/channel/39",
                      },
                    },
                    "name": "rock",
                    "number": 2,
                    "path": "channel",
                    "program": 0,
                    "programName": "rock",
                  }
                }
                key="2"
                onDelete={[Function]}
              >
                <div
                  className="row"
                >
                  <div
                    className="col-2"
                  >
                    2
                  </div>
                  <div
                    className="col-3"
                  >
                    rock
                  </div>
                  <div
                    className="col-1"
                  >
                    <button
                      className="btn btn-primary"
                      onClick={[Function]}
                      type="submit"
                    >
                      Delete
                    </button>
                  </div>
                </div>
              </Channel>
              <Channel
                entity={
                  Object {
                    "_links": Object {
                      "channel": Object {
                        "href": "http://localhost/api/rest/channel/40",
                      },
                      "performance": Object {
                        "href": "http://localhost/api/rest/channel/40/performance",
                      },
                      "self": Object {
                        "href": "http://localhost/api/rest/channel/40",
                      },
                    },
                    "name": "strings",
                    "number": 3,
                    "path": "channel",
                    "program": 0,
                    "programName": "strings",
                  }
                }
                key="3"
                onDelete={[Function]}
              >
                <div
                  className="row"
                >
                  <div
                    className="col-2"
                  >
                    3
                  </div>
                  <div
                    className="col-3"
                  >
                    strings
                  </div>
                  <div
                    className="col-1"
                  >
                    <button
                      className="btn btn-primary"
                      onClick={[Function]}
                      type="submit"
                    >
                      Delete
                    </button>
                  </div>
                </div>
              </Channel>
            </div>
          </div>
        </ChannelList>
      </div>
    </ChannelsContainer>
  </Connect(ChannelsContainer)>
</Provider>
`;
